numpy/core/arrayprint.py

```diff
@@ -638,74 +658,83 @@ def fillFormat(self, data):
-            max_str_len = signpos + int_len + 1 + precision
+            self.exp_size = None
+            max_str_len = signpos + int_len + 1 + sigfig
 
             if any(special):
                 neginf = self.sign != '-' or any(data[hasinf] < 0)
                 nanlen = len(_format_options['nanstr'])
                 inflen = len(_format_options['infstr']) + neginf
                 max_str_len = max(max_str_len, nanlen, inflen)
 
-            conversion = '' if self.sign == '-' else self.sign
-            format = '%#' + conversion + '%d.%df' % (max_str_len, precision)
-
+        conversion = '' if self.sign == '-' else self.sign
+        self.elem_format = self._makeFormatter(self.exp_format, conversion,
+                                               max_str_len, sigfig)
         self.special_fmt = '%%%ds' % (max_str_len,)
-        self.format = format
+
+    def _makeFormatter(self, exp_mode, conversion, max_str_len, sigfig):
+        fmt = "%#{}{}.{}{}".format(conversion, max_str_len, sigfig,
+                                   "e" if exp_mode else "f")
+        return lambda x: fmt % x
 
     def __call__(self, x, strip_zeros=True):
         with errstate(invalid='ignore'):
             if isnan(x):
                 nan_str = _format_options['nanstr']
-                if self.sign == '+':
-                    return self.special_fmt % ('+' + nan_str,)
-                else:
-                    return self.special_fmt % (nan_str,)
+                sign = '+' if self.sign == '+' else ''
+                return self.special_fmt % (sign + nan_str,)
             elif isinf(x):
                 inf_str = _format_options['infstr']
-                if x > 0:
-                    if self.sign == '+':
-                        return self.special_fmt % ('+' + inf_str,)
-                    else:
-                        return self.special_fmt % (inf_str,)
-                else:
-                    return self.special_fmt % ('-' + inf_str,)
-
-        s = self.format % x
-        if self.large_exponent:
-            # 3-digit exponent
-            expsign = s[-3]
-            if expsign == '+' or expsign == '-':
-                s = s[1:-2] + '0' + s[-2:]
-        elif self.exp_format:
-            # 2-digit exponent
-            if s[-3] == '0':
-                s = ' ' + s[:-3] + s[-2:]
+                sign = '-' if x < 0 else '+' if self.sign == '+' else ''
+                return self.special_fmt % (sign + inf_str,)
+
+        s = self.elem_format(x)
+
+        if self.exp_format:
+            # pad exponent with zeros, up to exp_size 0s.
+            frac_str, _, exp_str = s.partition('e')
+            expsign, expval = exp_str[0], exp_str[1:]
+            npad = self.exp_size -len(expval)
+            s = frac_str[npad:] + 'e' + expsign + '0'*npad + expval
         elif strip_zeros:
             z = s.rstrip('0')
             s = z + ' '*(len(s)-len(z))
         return s
 
+
+class LongFloatFormat(FloatFormat):
```

Comment:
I'd be inclined to make the base class `FloatingFormatter` to match `np.floating`, and use a derived class for `FloatFormatter` too.

Reply:
will do